File: mario/mta2/lib/mta.rb
require_relative 'single_line'

class MTA
	attr_reader :start_station, :finish_station,
				:counter, :stations

	NICE_OUTPUT =  {:n_line => "N-Line",
					:l_line => "L-Line",
					:six_line => "6-Line"
	               }

	def initialize(start_line, start_station, finish_line, finish_station)
		@start_line, @start_station = start_line, start_station
		@finish_line, @finish_station = finish_line, finish_station
	    @counter, @stations = 0, [] # Counter for the number of stations travelled
		journey
	end

	def start_line
		NICE_OUTPUT[@start_line]
	end

	def finish_line
		NICE_OUTPUT[@finish_line]
	end

	private

	def journey
		# if trip is on a single line
		if @start_line == @finish_line
			trip_with_no_transfers
		elsif @start_station == "Union Square"
			@start_line = @finish_line
			trip_with_no_transfers
		# else trip is on multiple lines
		else
			trip_with_transfers
		end
	end

	def trip_with_no_transfers
		s = SingleLine.new(@start_line, @start_station, @finish_station)
		@counter= s.counter
		@stations << s.stations
	end

	def trip_with_transfers
		s1 = SingleLine.new(@start_line, @start_station, "Union Square")
		s2 = SingleLine.new(@finish_line, "Union Square", @finish_station)
		@counter = s1.counter + s2.counter
		@stations << s1.stations << s2.stations
	end
end

File: mario/mta.js
// Third version of MTA, first version in Javascript

const LINES = {
    n_line : ["Time Square", "34th", "28th", "23rd", "Union Square", "8th"],
    l_line : ["8th", "6th", "Union Square", "3rd", "1st"],
    six_line : ["Grand Central", "33rd", "28th", "23rd", "Union Square", "Astor Place"]
}

singleLine = {
  line : "n_line",
  startStation : "28th",
  finishStation : "8th",

  startIndex : function() {
    return LINES[this.line].indexOf(this.startStation)
  },

  finishIndex : function() {
    return LINES[this.line].indexOf(this.finishStation)
  },

  counter : function() {
    return Math.abs(this.finishIndex() - this.startIndex());
  },

  stations : function() {
    if (this.finishIndex() === this.startIndex()) {
      return [LINES[this.line][this.startIndex]]
    } else if (this.startIndex() < this.finishIndex()) {
      return LINES[this.line].slice(this.startIndex(), this.finishIndex() + 1)
    } else {
      return (LINES[this.line].slice(this.finishIndex(), this.startIndex() + 1)).reverse()
    }
  }
}

mta = {
  startLine : "n_line",
  finishLine : "six_line",
  startStation : "28th",
  finishStation : "Grand Central",

  journey : function () {
    noLineOrStopError = this.errorChecker()

    if (noLineOrStopError ) {
      return noLineOrStopError 
    }


    if (this.startLine === this.finishLine) {
      return this.tripWithNoTransfers()
    } else {
      return this.tripWithTransfers()
    }
  },

  tripWithNoTransfers : function() {
    singleLine.line = this.startLine;
    singleLine.startStation = this.startStation;
    singleLine.finishStation = this.finishStation;
    output = []
    output.push("The number of stops is " + singleLine.counter());
    output.push("The stations are " + singleLine.stations());
    return output.join('\n')
  },

  tripWithTransfers : function() {
    singleLine.line = this.startLine;
    singleLine.startStation = this.startStation;
    singleLine.finishStation = "Union Square";
    counter = singleLine.counter();
    stations1 = singleLine.stations();

    singleLine.line = this.finishLine;
    singleLine.startStation = "Union Square";
    singleLine.finishStation = this.finishStation;
    counter += singleLine.counter();
    stations2 = singleLine.stations();

    output = []
    output.push("The number of stops is " + counter);
    output.push("The stop on " + this.startLine + " are " + stations1);
    output.push("Transfer at Union Square.");
    output.push("Then, the stop on " + this.finishLine + " are " + stations2);
    return output.join('\n')
  },

  errorChecker : function() {
    if (!(this.startLine in LINES)) {
      return "Please check your start line."
    } else if (!(this.finishLine in LINES)) {
      return "Please check your finish line."
    } else if (LINES[this.startLine].indexOf(this.startStation) === -1) {
      return "Please check your start station."
    } else if (LINES[this.finishLine].indexOf(this.startStation) === -1) {
      return "Please check your finish station."
    } else {
      return false
    }
  }
}

console.log(mta.journey());